### PR TITLE
Move actor-init-sparql's query engine contract types into the new package @comunica/types

### PR DESCRIPTION
**Pull request by jacoscaz**
This PR moves actor-init-sparql's contract types, i.e. the types that constitute the query engine contract of Comunica, into a separate types-only package `@comunica/types`. 

The goal of this new package is to play a similar role to `@types/rdf-js`, that is to provide shared typings without an explicit dependency on `@comunica/actor-init-sparql` and all of its dependencies.

The `@comunica/types` package should help with the following issues:

- https://github.com/beautifulinteractions/node-quadstore/issues/129
- https://github.com/comunica/comunica/issues/780

~~Something that could be improved is doing away with the re-export of the `IBindings` interface as the `Bindings` type in https://github.com/jacoscaz/comunica/blob/package/types/packages/bus-query-operation/lib/Bindings.ts#L10 and referencing the `IBindings` interface directly throughout the project. @rubensworks what do you think?~~ Actually, this turned out to be easier than I expected, so I just went ahead and did it.

### CODE REVIEW

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (rubensworks) [inline packages/bus-query-operation/lib/ActorQueryOperation.ts:187]: As to not require making this a breaking change, I would re-export the imported types here.

File: packages/bus-query-operation/lib/Bindings.ts
Comment (rubensworks) [inline packages/bus-query-operation/lib/Bindings.ts:16]: Also re-export this and following.

File: packages/bus-query-operation/package.json
Comment (rubensworks) [inline packages/bus-query-operation/package.json:45]: This would actually have to be an actual dependency. Also for all other packages (that use these types) for that matter. Strange that the linter didn't trip over this, as I thought I defined strict linter rules for this...

File: packages/core/lib/Actor.ts
Comment (rubensworks) [inline packages/core/lib/Actor.ts:7]: Ah, here you re-export, great :-)

File: packages/types/index.ts
Comment (rubensworks) [inline packages/types/index.ts:6]: Just wondering if we want to split this up into separate files, not sure yet...
In any case, we can still do this in a non-breaking future change if needed.

File: packages/types/index.ts
Comment (rubensworks) [inline packages/types/index.ts:14]: Any reason why this is prefixed with `I`, while the old type wasn't?
(Also applicable for other types in this file)

If no particular reason, I'd keep the type names exactly the same as before, as to not confuse users.

File: packages/types/package.json
Comment (rubensworks) [inline packages/types/package.json:5]: Or just removing this line would also be ok :-)

File: packages/types/package.json
Comment (rubensworks) [inline packages/types/package.json:34]: All of these would have to become actual dependencies based on how the types are defined now.

This makes me wonder if we really want this, or if we somehow want to define some more minimal interfaces of these packages internally, without having our types package blow up in number of interfaces.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Great! See my comments inline.

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/ActorQueryOperation.ts:187]: Ah yes, I will put those back in. I had dropped them to get the TS compiler to help me refactor stuff internally but not putting them back would likely break other packages depending on Comunica.

File: packages/bus-query-operation/lib/Bindings.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/Bindings.ts:16]: Will do.

File: packages/bus-query-operation/package.json
Comment (jacoscaz) [reply packages/bus-query-operation/package.json:45]: The general issue of errors not being 100% deterministic across environments is an annoying aspect of TS and TS linting, at least insofar as my limited experience goes. Will do.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts:6]: No strong opinion on this one.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts:14]: I think `IBindings` is a clearer indicator of what that is - a TS typing - and I think the `@comunica/types` could benefit from the additional clarity (given that most other types and interfaces do start with `I`). Same applies to `IBindingsStream`.

File: packages/types/package.json
Comment (jacoscaz) [reply packages/types/package.json:5]: Will do!

File: packages/types/package.json
Comment (jacoscaz) [reply packages/types/package.json:34]: I share this concern, although I have no idea of which percentage of these modules' typings we're actually using. I'll try to have a go at this and report back here.

File: packages/types/index.ts
Comment (rubensworks) [reply packages/types/index.ts:14]: We currently follow the convention that `interface`'s start with `I`, and types don't start with `I`.
I'm open to changing this, but then we'd have to make this consistent across the whole codebase, and make this a breaking change.

Comment (jacoscaz) [reply packages/types/index.ts:14]: Ah, you're right! My bad. Will drop the `I` from types. Apologies for the latency, I'm currently on a business trip and have very little time for anything else but the tests we're carrying out. At least the facility is nice!

Comment (rubensworks) [reply packages/types/index.ts:14]: No worries, there's no hurry on my end. Have fun there!

File: packages/types/package.json
Comment (jacoscaz) [reply packages/types/package.json]: Done!

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/ActorQueryOperation.ts:194]: This should be resolved now. I've opted to rename types using the `T` prefix instead of the `I` prefix in the new `@comunica/types` package (feels more coherent) but re-exported them without any prefix to maintain backward compatibility.

File: packages/bus-query-operation/lib/Bindings.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/Bindings.ts:16]: Further to my comments on Gitter, I gave up and simply re-added and re-exported the old `Bindings` and `BindingsStream` types while using the new `TBindings` and `TBindingsStream` ones throughout the module. They are effectively identical and the TS compiler is able to figure out that they are interchangeable. This likely comes at a cost in terms of compilation performance but I haven't seen tangible differences in compilation times, so I went with this workaround hoping that the night will bring some new idea as to how to approach this.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts:6]: Actually, I would rather keep them in the same file until the latter becomes exceedingly long and unmaintainable.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts]: See my comments above. I've replaced the `I` with `T` in the `@comunica/types` package as that felt like the most coherent solution but kept the original names when re-exporting types in packages like `@comunica/bus-query-operation` to maintain backward compatibility.

File: packages/bus-query-operation/package.json
Comment (jacoscaz) [reply packages/bus-query-operation/package.json:47]: Done!

File: packages/types/package.json
Comment (jacoscaz) [reply packages/types/package.json:33]: Gave a go at this one but found that Comunica is using quite a lot of both `AsyncIterator` and Immutable's `Map`, which would require quite a bit of typings duplication. Any ideas?

File: packages/types/package.json
Comment (rubensworks) [inline packages/types/package.json:34]:
```suggestion
    "@types/rdf-js": "*",
```

File: packages/types/package.json
Comment (rubensworks) [reply packages/types/package.json:33]: Yeah, I was afraid of that. Let's just keep it as-is then.

File: packages/types/index.ts
Comment (rubensworks) [reply packages/types/index.ts:6]: Sounds good to me.

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (rubensworks) [reply packages/bus-query-operation/lib/ActorQueryOperation.ts:194]: > I've opted to rename types using the T prefix instead of the I prefix in the new @comunica/types package

While I'm not opposed to the convention of prefixing types with `T`, such a change will have far-stretching consequences. If we want to be consistent, we'd have to prefix _all_ types in this repo with `T`.
Also, this may confuse users of the current types (without `T`).

For these reasons, I would suggest omitting the `T` prefix.

File: packages/bus-query-operation/lib/Bindings.ts
Comment (rubensworks) [reply packages/bus-query-operation/lib/Bindings.ts:16]: Did you experience any problems with just re-exporting those from the `types` package?
If not, I would propose to do that. Otherwise we'd have identical type and interface definitions in two different places, which will certainly cause maintainability problems in the future. (I would also expect jest's coverage requirement to trip over this)

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/ActorQueryOperation.ts:194]: No objections on my end. I thought that, given we're adding a new package, it might be a good idea to adopt a less confusing (IMHO) naming convention right from the start. But, I do understand the argument for consistency. I'll rename without the `T`.

File: packages/bus-query-operation/lib/Bindings.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/Bindings.ts:16]: Yes, once I drop the `T` and try to re-export the `Bindings` type as imported from `@comunica/types`, the TS compiler starts complaining about duplicate exports. I don't really know how to get out of this. Will test some more later on.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts]: I guess you can resolve this as we're also discussing the renaming issue in [the thread above](https://github.com/comunica/comunica/pull/784#discussion_r587715184).

File: packages/bus-query-operation/lib/ActorQueryOperation.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/ActorQueryOperation.ts:194]: Done!

File: packages/bus-query-operation/lib/Bindings.ts
Comment (jacoscaz) [reply packages/bus-query-operation/lib/Bindings.ts:16]: I might have found something that works. It's a bit ugly but it should prevent the maintainability issues you mentioned. The gist of it is as follows:

```js
import type { Bindings as _Bindings } from '@comunica/types';
export type Bindings = _Bindings;
function Bindings(hash) { /* ... */ }
```

The compiler seems happy and I've thus switched to this approach in my last commit. Done?

File: .eslintrc.js
Comment (rubensworks) [inline .eslintrc.js:111]: Was this disabled intentionally?

File: packages/actor-abstract-path/lib/ActorAbstractPath.ts
Comment (rubensworks) [inline packages/actor-abstract-path/lib/ActorAbstractPath.ts:208]: This can probably go if the next diff is undone.

File: packages/types/index.ts
Comment (rubensworks) [inline packages/types/index.ts:145]: I suspect some recent changes were lost. E.g. https://github.com/comunica/comunica/commit/12b9ee3e8e5bc2d0fadd662a3d6aeef838b87619#diff-cfc7a9b5296e9c5c4950a2fde534bebbe7053cc63ae511026bfa251f9e56a1b6

File: packages/types/index.ts
Comment (rubensworks) [inline packages/types/index.ts:159]: Since this is a brand new interface, we may consider giving this a more generic name. Perhaps something like `IQueryEngine`?
What do you think?

File: packages/types/index.ts
Comment (rubensworks) [inline packages/types/index.ts:97]:
```suggestion
export type IActorQueryOperationOutput =
```

I'm aware this is not an interface, but I would do this to be better in line with the currently existing interface/type (so that developers only have to change the import target, not the rest of their code).

We may want to add a `TODO` comment above this to properly rename upon our next major update. Perhaps we could then even use a name such as `QueryOutput`.

**Review by rubensworks — CHANGES_REQUESTED** [review]
Just some minor nits. Should be good to go after these suggestions.

File: .eslintrc.js
Comment (jacoscaz) [reply .eslintrc.js]: I just removed this line. Something in my workflow had altered param indentation in a few places but I got tricked into thinking I had inherited such indentation from `master`.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts]: Agreed! And done.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts]: Done!

**Comment by jacoscaz** [comment]
Something apparently unrelated to this PR but that comes up at least twice a day: running `yarn run test` or `yarn run test-ci` often fails with the following error:

```
 FAIL  packages/actor-rdf-resolve-quad-pattern-hypermedia/test/LinkedRdfSourcesAsyncRdfIterator-test.ts
  ● LinkedRdfSourcesAsyncRdfIterator › A LinkedRdfSourcesAsyncRdfIterator instance › handles metadata event emitted before the end event

    Unhandled error. (TypeError: require(...) is not a function



      at Function.toNodeReadable (packages/bus-http/lib/ActorHttp.ts:376:32)
      at ActorRdfDereferenceHttpParse.run (packages/actor-rdf-dereference-http-parse/lib/ActorRdfDereferenceHttpParseBase.ts:1480:53)
      at MediatedLinkedRdfSourcesAsyncRdfIterator.getSource (packages/actor-rdf-resolve-quad-pattern-hypermedia/lib/MediatedLinkedRdfSourcesAsyncRdfIterator.ts:905:32))Error [ERR_UNHANDLED_ERROR]: Unhandled error. (TypeError: require(...) is not a function
      at Function.toNodeReadable (packages/bus-http/lib/ActorHttp.ts:376:32)
      at ActorRdfDereferenceHttpParse.run (packages/actor-rdf-dereference-http-parse/lib/ActorRdfDereferenceHttpParseBase.ts:1480:53)
      at MediatedLinkedRdfSourcesAsyncRdfIterator.getSource (packages/actor-rdf-resolve-quad-pattern-hypermedia/lib/MediatedLinkedRdfSourcesAsyncRdfIterator.ts:905:32))
      at SimpleTransformIterator.<anonymous> (packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:3039:19)
      at MediatedLinkedRdfSourcesAsyncRdfIterator.destinationEmitError (node_modules/asynciterator/dist/asynciterator.cjs:1067:23)
      at node_modules/asynciterator/dist/asynciterator.cjs:177:26
      at MediatedLinkedRdfSourcesAsyncRdfIterator._destroy (node_modules/asynciterator/dist/asynciterator.cjs:868:9)
      at MediatedLinkedRdfSourcesAsyncRdfIterator.destroy (node_modules/asynciterator/dist/asynciterator.cjs:174:18)
      at packages/actor-rdf-resolve-quad-pattern-hypermedia/lib/LinkedRdfSourcesAsyncRdfIterator.ts:200:30
```

Whenever this happens, running the test suite again always succeeds without errors. I don't think it's related to something I've done but I thought I'd mention it anyway, if only to keep track of it.

File: packages/actor-abstract-path/lib/ActorAbstractPath.ts
Comment (jacoscaz) [reply packages/actor-abstract-path/lib/ActorAbstractPath.ts]: Done!

**Comment by rubensworks** [comment]
> running yarn run test or yarn run test-ci often fails with the following error:

Hmm, I never seen that error before (locally or CI).
Any chance you're using a very recent (experimental?) Node version?
If not, feel free to open an issue for this, just in case others would also run into this problem.

**Comment by jacoscaz** [comment]
I generally tend to stick with the LTS branch, I'm currently using `v14.15.4`. Will open an issue.

File: packages/types/index.ts
Comment (jacoscaz) [reply packages/types/index.ts]: I completely missed the updated version of this jsdoc comment in moving the `ActionContext` type to `@comunica/types` and then rebasing my branch onto `master`. I've now updated the comment in `@comunica/types`  to match the version in `master` and modified the latter to add the `@deprecated` flag. Should be done.

**Comment by jacoscaz** [comment]
Perhaps I should squash all of these commits into a single one after your review? It'd make the commit history a bit tidier.

**Comment by rubensworks** [comment]
Oh, that's ok, I can squash here via the GitHub UI before merging :-)

File: packages/actor-query-operation-leftjoin-left-deep/lib/ActorQueryOperationLeftJoinLeftDeep.ts
Comment (rubensworks) [inline packages/actor-query-operation-leftjoin-left-deep/lib/ActorQueryOperationLeftJoinLeftDeep.ts:35]: I assume this line can also go?

Comment (jacoscaz) [reply packages/actor-query-operation-leftjoin-left-deep/lib/ActorQueryOperationLeftJoinLeftDeep.ts]: Done!

**Comment by rubensworks** [comment]
Perfect, thanks!

Any urgency on a release from your end?

**Comment by jacoscaz** [comment]
It'd be nice to be able to reference the new `@comunica/types` with the next release of `quadstore` and `quadstore-comunica` so as to remove the duplicated typings but there's plenty to do WRT to the query spec, so no rush at all.

**Comment by jacoscaz** [comment]
Thank you for merging this!

**Pull request merged**
